api/v1alpha1/resourcepipeline_types.go

```diff
@@ -0,0 +1,160 @@
+// SPDX-FileCopyrightText: 2022 SAP SE or an SAP affiliate company and Open Component Model contributors.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package v1alpha1
+
+import (
+	"strings"
+	"time"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ResourcePipelineSpec defines the desired state of ResourcePipeline
+type ResourcePipelineSpec struct {
+	// +required
+	Interval metav1.Duration `json:"interval"`
+
+	// +optional
+	Suspend bool `json:"suspend,omitempty"`
+
+	// +optional
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+
+	// +required
+	SourceRef ObjectReference `json:"sourceRef"`
+
+	// +optional
+	Parameters *apiextensionsv1.JSON `json:"parameters,omitempty"`
+
+	// +optional
+	PipelineSpec *PipelineSpec `json:"pipelineSpec,omitempty"`
+}
+
+// ResourcePipelineSource defines the component version and resource
+// which will be processed by the pipeline.
+type ResourcePipelineSource struct {
+	// +required
+	Name string `json:"name"`
+
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// +required
+	Resource string `json:"resource"`
+}
+
+// PipelineSpec holds the steps that constitute the pipeline.
+type PipelineSpec struct {
+	// +required
+	Steps []WasmStep `json:"steps"`
+}
+
+// DeliverySpec holds a set of targets onto which the pipeline output will be deployed.
+type DeliverySpec struct {
+	// +required
+	Targets []WasmStep `json:"targets"`
+}
+
+// WasmStep defines the name version and location of a wasm module that is stored// in an ocm component. The format of the module name must be <component-name>:<component-version>@<resource-name>. Optionally a registry address can be specified.
+type WasmStep struct {
+	// +required
+	Name string `json:"name"`
+
+	// +kubebuilder:example="ocm.software/modules:v1.0.0@kustomizer"
+	// +kubebuilder:validation:Pattern="^([A-Za-z0-9\\.\\/]+):(v[0-9\\.\\-a-z]+)@([a-z]+)$"
+	// +required
+	Module string `json:"module"`
+
+	// +kubebuilder:default="ghcr.io/open-component-model"
+	// +optional
+	Registry string `json:"registry,omitempty"`
+
+	// +optional
+	Values *apiextensionsv1.JSON `json:"values,omitempty"`
+}
+
+func (w WasmStep) GetComponent() string {
+	return strings.Split(w.Module, ":")[0]
+}
+
+func (w WasmStep) GetComponentVersion() string {
+	p1 := strings.Split(w.Module, ":")[1]
+	return strings.Split(p1, "@")[0]
+}
+
+func (w WasmStep) GetResource() string {
+	return strings.Split(w.Module, "@")[1]
+}
```

Comment:
So this is interesting. This means that any module can be executed on any component version where ever that module is located at. That also means that potentially you would have to configure access to that module. Shouldn't the module that is currently being processed contain the wasm binaries that it wants to execute?

I'm just wondering about that the potential outside module is compromised and we would have no way of knowing because this just tries to fetch it as is.

Reply:
We need to add signature verification here. I think that could be a follow up? Or I can put it in this PR.

It should be possible to pull from any Component containing modules not just the current component.

Reply:
Interesting. What if you make a typo in the module's name, and it suddenly pulls from a different module than you would expect?

Reply:
That could happen with any kind of remote resource; it you want to lock down what registries or components you pull from then you do that via a policy that is written to validate the ResourcePipeline (gatekeeper or kyverno etc...).

Reply:
Okay then :)